rules/do-you-use-loggermessage-in-net.md/rule.md

```diff
@@ -0,0 +1,118 @@
+
+2. **Efficient Message Parsing**: Logger extension methods parse the log message template each time a log entry is made. In contrast, LoggerMessage only parses the template once when the message is defined, leading to fewer computations and faster execution during logging.
+
+3. **Structured Logging**: With LoggerMessage, log templates can include placeholders that are filled with specific log data, such as Item or DateTime. This makes it easier to integrate structured logging with logging frameworks like Serilog, which can enrich logs with meaningful event names.
+
+## How to use LoggerMessage
+
+To use the LoggerMessage pattern, you define a static Action delegate for logging a message. The LoggerMessage.Define method is used to create these delegates with parameters such as:
+
+- Log Level (e.g., LogLevel.Information)
+- Event ID (for tracking specific events)
+- Message Template (a named format string with placeholders)
+
+For example:
+
+Each log message is represented as an Action delegate.
+
+```csharp
+private static readonly Action<ILogger, WorkItem, Exception> s_processingPriorityItem =
+    LoggerMessage.Define<WorkItem>(
+        LogLevel.Information,
+        new EventId(1, nameof(PriorityItemProcessed)),
+        "Processing priority item: {Item}");
+
+private static readonly Action<ILogger, Exception> s_failedToProcessWorkItem = LoggerMessage.Define<Exception>(
+    LogLevel.Critical,
+    new EventId(13, nameof(FailedToProcessWorkItem)),
+    "Epic failure processing item!");
+```
+
+This delegate is then invoked using strongly typed extension methods like PriorityItemProcessed, making it simple to log information in a structured manner with minimal performance impact.
+
+```csharp
+public static void PriorityItemProcessed(
+    this ILogger logger, WorkItem workItem) =>
+    s_processingPriorityItem(logger, workItem, default!);
+
+public static void FailedToProcessWorkItem(
+    this ILogger logger, Exception  ex) =>
+    s_failedToProcessWorkItem(logger, ex);
+```
+
+Usage example:
+
+```csharp
+protected override async Task ExecuteAsync(
+    CancellationToken stoppingToken)
+{
+    using (IDisposable? scope = logger.ProcessingWorkScope(DateTime.Now))
+    {
+        while (!stoppingToken.IsCancellationRequested)
+        {
+            try
+            {
+                WorkItem? nextItem = priorityQueue.ProcessNextHighestPriority();
+
+                if (nextItem is not null)
+                {
+                    logger.PriorityItemProcessed(nextItem);
+                }
+            }
+            catch (Exception ex)
+            {
+                logger.FailedToProcessWorkItem(ex);
+            }
+
+            await Task.Delay(1_000, stoppingToken);
+        }
+    }
+}
+
+```
+
+## Optimizing further with ILogger.IsEnabled
```

Comment:
Double check, but I think the code generated by the logger message templates does this for you.

Reply:
@tomek-i - did you look into this?